File: sdk/python/aicomply_logvault.py
"""
AIComply LogVault SDK — Python
EU AI Act Art. 12 — Automatic logging for high-risk AI systems

Usage:
    from aicomply_logvault import LogVault

    lv = LogVault(api_key="your-key", ai_system_id="uuid")

    # Manual logging
    lv.log_inference(
        prompt="Analizza questo CV",
        output="Score: 87/100",
        model_id="gpt-4o",
        latency_ms=320
    )

    # OpenAI auto-interceptor
    import openai
    lv.patch_openai(openai)

    # LangChain callback
    from langchain.callbacks import CallbackManager
    manager = CallbackManager([lv.langchain_callback()])
"""

import hashlib
import json
import time
import threading
import requests
from datetime import datetime
from typing import Optional, Dict, Any, List
from queue import Queue, Empty


AICOMPLY_ENDPOINT = "https://aicomply-omega.vercel.app/api/logvault/ingest"
SDK_VERSION = "1.0.0"


class LogVault:
    """
    AIComply LogVault — Art. 12 EU AI Act compliance logger.
    Thread-safe, async-friendly, with automatic batch flushing.
    """

    def __init__(
        self,
        api_key: str,
        ai_system_id: Optional[str] = None,
        endpoint: str = AICOMPLY_ENDPOINT,
        batch_size: int = 10,
        flush_interval: float = 5.0,
        privacy_mode: bool = True,   # Se True, invia solo hash (mai testo in chiaro)
        enabled: bool = True,
    ):
        self.api_key = api_key
        self.ai_system_id = ai_system_id
        self.endpoint = endpoint
        self.batch_size = batch_size
        self.flush_interval = flush_interval
        self.privacy_mode = privacy_mode
        self.enabled = enabled

        self._queue: Queue = Queue()
        self._lock = threading.Lock()
        self._session = requests.Session()
        self._session.headers.update({
            "X-AIComply-Key": api_key,
            "X-AIComply-SDK-Version": SDK_VERSION,
            "Content-Type": "application/json",
        })

        # Avvia flush thread in background
        if enabled:
            self._flush_thread = threading.Thread(
                target=self._flush_loop,
                daemon=True,
                name="aicomply-logvault-flush"
            )
            self._flush_thread.start()

    # ─── Core logging ──────────────────────────────────────────────────────────

    def log_inference(
        self,
        prompt: str,
        output: str,
        model_id: str = "unknown",
        model_version: Optional[str] = None,
        latency_ms: Optional[int] = None,
        token_count: Optional[int] = None,
        within_guardrails: bool = True,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Logga un evento di inferenza AI."""
        if not self.enabled:
            return

        event = {
            "event_type": "inference",
            "model_id": model_id,
            "model_version": model_version,
            "latency_ms": latency_ms,
            "token_count": token_count,
            "within_guardrails": within_guardrails,
            "metadata": metadata or {},
        }

        if self.privacy_mode:
            # Invia solo hash SHA-256 — mai testo in chiaro
            event["prompt_hash"] = self._hash(prompt)
            event["output_hash"] = self._hash(output)
        else:
            event["prompt"] = prompt
            event["output"] = output
            event["prompt_hash"] = self._hash(prompt)
            event["output_hash"] = self._hash(output)

        if self.ai_system_id:
            event["ai_system_id"] = self.ai_system_id

        self._queue.put(event)

    def log_alert(
        self,
        reason: str,
        severity: str = "warning",
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Logga un alert di compliance."""
        if not self.enabled:
            return

        event = {
            "event_type": "alert",
            "flagged": True,
            "flag_reason": reason,
            "flag_severity": severity,
            "within_guardrails": False,
            "metadata": metadata or {},
        }
        if self.ai_system_id:
            event["ai_system_id"] = self.ai_system_id

        self._queue.put(event)

    def log_drift(
        self,
        metric: str,
        current_value: float,
        baseline_value: float,
        threshold: float,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Logga un evento di data/model drift."""
        if not self.enabled:
            return

        deviation_pct = abs(current_value - baseline_value) / max(abs(baseline_value), 1e-9) * 100
        event = {
            "event_type": "drift",
            "flagged": deviation_pct > threshold,
            "flag_reason": f"Drift detected: {metric} = {current_value:.3f} (baseline: {baseline_value:.3f}, deviation: {deviation_pct:.1f}%)",
            "flag_severity": "critical" if deviation_pct > threshold * 2 else "warning",
            "within_guardrails": deviation_pct <= threshold,
            "metadata": {
                **(metadata or {}),
                "metric": metric,
                "current": current_value,
                "baseline": baseline_value,
                "threshold": threshold,
                "deviation_pct": round(deviation_pct, 2),
            },
        }
        if self.ai_system_id:
            event["ai_system_id"] = self.ai_system_id

        self._queue.put(event)

    # ─── OpenAI interceptor ────────────────────────────────────────────────────

    def patch_openai(self, openai_module) -> None:
        """
        Intercetta automaticamente le chiamate OpenAI.

        Uso:
            import openai
            lv.patch_openai(openai)
            # Da questo momento tutte le chat.completions.create vengono loggati
        """
        original_create = openai_module.chat.completions.create
        lv = self

        def patched_create(*args, **kwargs):
            start = time.time()
            response = original_create(*args, **kwargs)
            latency = int((time.time() - start) * 1000)

            try:
                messages = kwargs.get("messages", [])
                prompt = messages[-1].get("content", "") if messages else ""
                output = response.choices[0].message.content if response.choices else ""
                usage = response.usage
                lv.log_inference(
                    prompt=prompt,
                    output=output,
                    model_id=kwargs.get("model", "openai-unknown"),
                    latency_ms=latency,
                    token_count=usage.total_tokens if usage else None,
                    metadata={"openai_id": response.id},
                )
            except Exception:
                pass  # Non interrompere mai il flusso principale

            return response

        openai_module.chat.completions.create = patched_create
        print(f"[AIComply LogVault] OpenAI interceptor active (privacy_mode={self.privacy_mode})")

    # ─── LangChain callback ────────────────────────────────────────────────────

    def langchain_callback(self):
        """
        Restituisce un LangChain BaseCallbackHandler per logging automatico.

        Uso:
            from langchain_core.callbacks import CallbackManager
            chain = MyChain(callbacks=[lv.langchain_callback()])
        """
        try:
            from langchain_core.callbacks import BaseCallbackHandler

            lv = self

            class AIComplyCallback(BaseCallbackHandler):
                def on_llm_start(self, serialized, prompts, **kwargs):
                    self._start_time = time.time()
                    self._prompts = prompts

                def on_llm_end(self, response, **kwargs):
                    latency = int((time.time() - getattr(self, "_start_time", time.time())) * 1000)
                    prompt = self._prompts[0] if getattr(self, "_prompts", []) else ""
                    output = response.generations[0][0].text if response.generations else ""
                    lv.log_inference(
                        prompt=prompt,
                        output=output,
                        model_id="langchain",
                        latency_ms=latency,
                    )

            return AIComplyCallback()
        except ImportError:
            raise ImportError("langchain_core required: pip install langchain-core")

    # ─── Flush logic ───────────────────────────────────────────────────────────

    def flush(self) -> Dict[str, Any]:
        """Forza il flush immediato della coda."""
        events: List[Dict] = []
        try:
            while True:
                events.append(self._queue.get_nowait())
                if len(events) >= self.batch_size:
                    break
        except Empty:
            pass

        if not events:
            return {"logged": 0}

        return self._send_batch(events)

    def _flush_loop(self) -> None:
        while True:
            time.sleep(self.flush_interval)
            self.flush()

    def _send_batch(self, events: List[Dict]) -> Dict[str, Any]:
        try:
            payload = {
                "api_key": self.api_key,
                "events": events,
            }
            resp = self._session.post(
                self.endpoint,
                json=payload,
                timeout=10,
            )
            resp.raise_for_status()
            return resp.json()
        except Exception as e:
            print(f"[AIComply LogVault] Flush error: {e}")
            return {"logged": 0, "error": str(e)}

    @staticmethod
    def _hash(text: str) -> str:
        return hashlib.sha256(text.encode("utf-8")).hexdigest()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.flush()


# ─── Decoratori helper ────────────────────────────────────────────────────────

def log_ai_call(lv: LogVault, model_id: str = "unknown"):
    """
    Decoratore per loggare automaticamente funzioni che chiamano AI.

    Uso:
        @log_ai_call(lv, model_id="gpt-4o")
        def mia_funzione_ai(prompt: str) -> str:
            return client.chat.completions.create(...)
    """
    def decorator(func):
        def wrapper(*args, **kwargs):
            start = time.time()
            result = func(*args, **kwargs)
            latency = int((time.time() - start) * 1000)
            prompt = str(args[0]) if args else str(kwargs)
            output = str(result)
            lv.log_inference(
                prompt=prompt,
                output=output,
                model_id=model_id,
                latency_ms=latency,
            )
            return result
        return wrapper
    return decorator

File: sdk/js/aicomply-logvault.ts
/**
 * AIComply LogVault SDK — TypeScript/JavaScript
 * EU AI Act Art. 12 — Automatic logging for high-risk AI systems
 *
 * Usage:
 *   import { LogVault } from "@aicomply/logvault";
 *
 *   const lv = new LogVault({ apiKey: "your-key", aiSystemId: "uuid" });
 *
 *   // Manual logging
 *   await lv.logInference({ prompt, output, modelId: "gpt-4o", latencyMs: 320 });
 *
 *   // OpenAI auto-interceptor
 *   import OpenAI from "openai";
 *   const openai = lv.wrapOpenAI(new OpenAI());
 *
 *   // Vercel AI SDK middleware
 *   export const runtime = lv.vercelMiddleware();
 */

import crypto from "crypto";

const AICOMPLY_ENDPOINT = "https://aicomply-omega.vercel.app/api/logvault/ingest";
const SDK_VERSION = "1.0.0";

// ─── Types ────────────────────────────────────────────────────────────────────

export type EventType = "inference" | "alert" | "drift" | "audit" | "error";
export type Severity  = "info" | "warning" | "critical";

export interface LogEvent {
  event_type:        EventType;
  ai_system_id?:     string;
  model_id?:         string;
  model_version?:    string;
  prompt?:           string;
  prompt_hash?:      string;
  output?:           string;
  output_hash?:      string;
  latency_ms?:       number;
  token_count?:      number;
  flagged?:          boolean;
  flag_reason?:      string;
  flag_severity?:    Severity;
  within_guardrails?: boolean;
  metadata?:         Record<string, unknown>;
}

export interface LogVaultConfig {
  apiKey:          string;
  aiSystemId?:     string;
  endpoint?:       string;
  batchSize?:      number;
  flushInterval?:  number;  // ms
  privacyMode?:    boolean; // default true — invia solo hash
  enabled?:        boolean;
}

export interface FlushResult {
  logged:   number;
  alerts:   { event_index: number; reason: string; severity: string }[] | null;
  chain_head?: string;
  error?:   string;
}

// ─── LogVault class ───────────────────────────────────────────────────────────

export class LogVault {
  private readonly config:    Required<LogVaultConfig>;
  private readonly queue:     LogEvent[] = [];
  private          timer:     ReturnType<typeof setInterval> | null = null;

  constructor(config: LogVaultConfig) {
    this.config = {
      endpoint:      AICOMPLY_ENDPOINT,
      batchSize:     10,
      flushInterval: 5000,
      privacyMode:   true,
      enabled:       true,
      aiSystemId:    "",
      ...config,
    };

    if (this.config.enabled) {
      this.timer = setInterval(() => this.flush(), this.config.flushInterval);
    }
  }

  // ─── Core logging ───────────────────────────────────────────────────────────

  async logInference(params: {
    prompt:         string;
    output:         string;
    modelId?:       string;
    modelVersion?:  string;
    latencyMs?:     number;
    tokenCount?:    number;
    withinGuardrails?: boolean;
    metadata?:      Record<string, unknown>;
  }): Promise<void> {
    if (!this.config.enabled) return;

    const event: LogEvent = {
      event_type:        "inference",
      model_id:          params.modelId,
      model_version:     params.modelVersion,
      latency_ms:        params.latencyMs,
      token_count:       params.tokenCount,
      within_guardrails: params.withinGuardrails ?? true,
      metadata:          params.metadata,
      prompt_hash:       this.hash(params.prompt),
      output_hash:       this.hash(params.output),
    };

    if (!this.config.privacyMode) {
      event.prompt = params.prompt;
      event.output = params.output;
    }

    if (this.config.aiSystemId) {
      event.ai_system_id = this.config.aiSystemId;
    }

    this.enqueue(event);
  }

  async logAlert(params: {
    reason:    string;
    severity?: Severity;
    metadata?: Record<string, unknown>;
  }): Promise<void> {
    if (!this.config.enabled) return;

    this.enqueue({
      event_type:        "alert",
      flagged:           true,
      flag_reason:       params.reason,
      flag_severity:     params.severity ?? "warning",
      within_guardrails: false,
      metadata:          params.metadata,
      ai_system_id:      this.config.aiSystemId || undefined,
    });
  }

  async logDrift(params: {
    metric:        string;
    currentValue:  number;
    baselineValue: number;
    threshold:     number;
    metadata?:     Record<string, unknown>;
  }): Promise<void> {
    if (!this.config.enabled) return;

    const deviation = Math.abs(params.currentValue - params.baselineValue)
      / Math.max(Math.abs(params.baselineValue), 1e-9) * 100;

    this.enqueue({
      event_type:        "drift",
      flagged:           deviation > params.threshold,
      flag_reason:       `Drift: ${params.metric} = ${params.currentValue.toFixed(3)} (baseline: ${params.baselineValue.toFixed(3)}, deviation: ${deviation.toFixed(1)}%)`,
      flag_severity:     deviation > params.threshold * 2 ? "critical" : "warning",
      within_guardrails: deviation <= params.threshold,
      metadata: {
        ...params.metadata,
        metric:       params.metric,
        current:      params.currentValue,
        baseline:     params.baselineValue,
        threshold:    params.threshold,
        deviation_pct: Math.round(deviation * 100) / 100,
      },
      ai_system_id: this.config.aiSystemId || undefined,
    });
  }

  // ─── OpenAI wrapper ─────────────────────────────────────────────────────────

  wrapOpenAI<T extends { chat: { completions: { create: (...args: unknown[]) => Promise<unknown> } } }>(client: T): T {
    const lv = this;
    const originalCreate = client.chat.completions.create.bind(client.chat.completions);

    client.chat.completions.create = async (...args: unknown[]) => {
      const start = Date.now();
      const response = await originalCreate(...args) as {
        choices?: Array<{ message?: { content?: string } }>;
        usage?: { total_tokens?: number };
        id?: string;
        model?: string;
      };
      const latencyMs = Date.now() - start;

      try {
        const params = args[0] as {
          messages?: Array<{ content?: string }>;
          model?: string;
        };
        const messages = params?.messages ?? [];
        const prompt = messages.at(-1)?.content ?? "";
        const output = response.choices?.[0]?.message?.content ?? "";

        await lv.logInference({
          prompt,
          output,
          modelId:    response.model ?? params?.model ?? "openai",
          latencyMs,
          tokenCount: response.usage?.total_tokens,
          metadata:   { openai_id: response.id },
        });
      } catch {
        // Mai bloccare il flusso principale
      }

      return response;
    };

    console.log(`[AIComply LogVault] OpenAI wrapper active (privacyMode=${this.config.privacyMode})`);
    return client;
  }

  // ─── Next.js / Edge middleware helper ───────────────────────────────────────

  nextMiddleware() {
    const lv = this;
    return async function logVaultMiddleware(
      req: { method: string; url: string; headers: Headers },
      next: () => Promise<Response>
    ): Promise<Response> {
      const start = Date.now();
      const response = await next();
      const latencyMs = Date.now() - start;

      // Logga solo rotte AI
      if (req.url.includes("/api/ai") || req.url.includes("/api/chat")) {
        await lv.logInference({
          prompt:   `${req.method} ${req.url}`,
          output:   `HTTP ${response.status}`,
          modelId:  "next-api",
          latencyMs,
        });
      }

      return response;
    };
  }

  // ─── Flush ──────────────────────────────────────────────────────────────────

  async flush(): Promise<FlushResult> {
    const events = this.queue.splice(0, this.config.batchSize);
    if (events.length === 0) return { logged: 0, alerts: null };

    try {
      const response = await fetch(this.config.endpoint, {
        method:  "POST",
        headers: {
          "Content-Type":           "application/json",
          "X-AIComply-Key":         this.config.apiKey,
          "X-AIComply-SDK-Version": SDK_VERSION,
        },
        body: JSON.stringify({ api_key: this.config.apiKey, events }),
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      return await response.json() as FlushResult;
    } catch (err) {
      // Re-accoda gli eventi non inviati
      this.queue.unshift(...events);
      return { logged: 0, alerts: null, error: String(err) };
    }
  }

  destroy(): void {
    if (this.timer) clearInterval(this.timer);
    this.flush().catch(() => {});
  }

  // ─── Utility ────────────────────────────────────────────────────────────────

  private enqueue(event: LogEvent): void {
    this.queue.push(event);
    if (this.queue.length >= this.config.batchSize) {
      this.flush().catch(() => {});
    }
  }

  private hash(text: string): string {
    return crypto.createHash("sha256").update(text, "utf8").digest("hex");
  }
}

// ─── Singleton helper ────────────────────────────────────────────────────────

let _instance: LogVault | null = null;

export function createLogVault(config: LogVaultConfig): LogVault {
  _instance = new LogVault(config);
  return _instance;
}

export function getLogVault(): LogVault {
  if (!_instance) throw new Error("LogVault not initialized — call createLogVault() first");
  return _instance;
}
